pgmpy/utils/utils.py

```diff
@@ -399,7 +401,19 @@ def preprocess_data(df):
                 f"Couldn't infer datatype of column: {col} from data. Try specifying the appropriate datatype to the column."
             )
 
-    logger.info(
-        f" Datatype (N=numerical, C=Categorical Unordered, O=Categorical Ordered) inferred from data: \n {dtypes}"
-    )
+    key = tuple(sorted((col, str(df[col].dtype)) for col in df.columns))
+
+    # Log only if this structure hasn't been logged before
+    if key not in _logged_keys:
+        logger.info(
+            f" Datatype (N=numerical, C=Categorical Unordered, O=Categorical Ordered) inferred from data: \n {dtypes}"
+        )
+        _logged_keys.add(key)
+
+    for col in df.columns:
+        if pd.api.types.is_integer_dtype(df[col]):
+            df[col] = df[col].astype("int")
+        elif pd.api.types.is_object_dtype(df[col]):
+            df[col] = df[col].astype("category")
```

Comment:
I am still not sure about Lines 413-417. Is this really required if the checks and conversion have already been done in the for loop in Line 385?